### PR TITLE
[AI] More work on economic policies

### PR DESCRIPTION
**Pull request by Grummel7**
I found some time to work on policies again...

### CODE REVIEW

File: default/python/AI/ResourcesAI.py [hunk 1]
```diff
@@ -94,6 +94,7 @@ def __init__(self):
             aistate.get_priority(PriorityType.RESOURCE_RESEARCH),
             aistate.get_priority(PriorityType.RESOURCE_INFLUENCE),
         )
+        self.bureaucracy = bureaucracy in fo.getEmpire().adoptedPolicies
```
Comment (Cjkjvfnby) [inline default/python/AI/ResourcesAI.py]: Maybe name it `has_bureaucracy`?

File: default/python/AI/ResourcesAI.py [hunk 2]
```diff
@@ -159,6 +147,24 @@ def bake_future_focus(self, pid, focus, update=True, force=False):
...
         return success
 
+    def check_avoid_change_due_to_bureaucracy(self, pinfo, focus, focus_gain, influence_gain) -> bool:
+        if not self.bureaucracy or focus == pinfo.current_focus:
+            return False  # No bureaucracy or no change
+        last_turn = fo.currentTurn() - 1
+        if pinfo.planet.LastTurnColonized == last_turn or pinfo.planet.LastTurnConquered == last_turn:
+            return False  # no penalty for newly settled/conquered planets
+        threshold = 3.0
+        if (
```
Comment (Cjkjvfnby) [inline default/python/AI/ResourcesAI.py]: You could use `return any()`

Maybe you could try to extract each group to the variable? 
Naming would be a challenge here.

```
not_enought_influence = (focus == INFLUENCE and influence_gain < focus_gain + threshold)
...focus... = (pinfo.current_focus == INFLUENCE and focus_gain < influence_gain + threshold)
```

File: default/python/AI/PolicyAI.py [hunk 3]
```diff
@@ -97,8 +109,22 @@ def add_planet(self, planet: fo.planet) -> None:
         elif stability >= 10:
             # 10 give a lot of bonuses, increases above 10 are less important
             stability = 11 + (stability - 10) / 2
+        if self.has_liberty and planet.focus == FocusType.FOCUS_RESEARCH:
+            self.adjust_liberty(planet, population)
         self.population_stability += population * stability * self.stability_scaling
 
+    def adjust_liberty(self, planet: fo.planet, population: float) -> None:
```
Comment (Cjkjvfnby) [inline default/python/AI/PolicyAI.py]: Maybe name it `add_liberty_bonus`?

File: default/python/AI/PolicyAI.py [hunk 4]
```diff
@@ -97,8 +109,22 @@ def add_planet(self, planet: fo.planet) -> None:
...
+        """adjust liberty research output for changed stability"""
+        # This one is relatively easy to calculate. Still it would be better to have an adjustMeterUpdate with
+        # modified stability...
+        current_stability = planet.currentMeterValue(fo.meterType.happiness)
+        target_stability = planet.currentMeterValue(fo.meterType.targetHappiness)
+        low = get_named_real("PLC_LIBERTY_MIN_STABILITY")
+        high = get_named_real("PLC_LIBERTY_MAX_STABILITY")
+        difference = max(low, min(high, target_stability)) - max(low, min(high, current_stability))
+        debug(f"adjust_liberty on {planet}: difference={difference}, population={population}")
+        self.research += difference * population * get_named_real("PLC_LIBERTY_RESEARCH_BONUS_SCALING")
```
Comment (Cjkjvfnby) [inline default/python/AI/PolicyAI.py]: How about moving the mutation of self out of this function?

```
self.research += self.adjust_liberty(
```

**Review by Cjkjvfnby — COMMENTED** [review]
Left some comments

File: default/python/AI/ResourcesAI.py (hunk 2 above)
Comment (Grummel7) [reply default/python/AI/ResourcesAI.py]: This part was actually not working as intended, so I had to partially rewrite it.
The whole module is on my list of things that should be rewritten. E.g. it still assumes that changing e.g. from research to industry focus only decreases RP and increases PP, ignoring the effects of stability.

File: default/python/AI/ResourcesAI.py [hunk 5]
```diff
@@ -147,24 +141,40 @@ def bake_future_focus(self, pid, focus, update=True, force=False):
...
+        """
+        Calculate value of the production output with the given production focus.
+        Returns 0 of focus is not INDUSTRY, RESEARCH or INFLUENCE, unless focus is pinfo.current_focus
+        """
+        if focus not in (INDUSTRY, RESEARCH, INFLUENCE, pinfo.current_focus):
+            return 0.0
+        # Note that changes in stability may affect other values, so there could be cases where a planet
+        # produces more PP with research focus than with industry focus. This is not correctly evaluated yet,
+        # but let's sum up all three anyway. The whole module should be refactored or possible re-written...
+        return (
```
Comment (Cjkjvfnby) [inline default/python/AI/ResourcesAI.py:172]: Maybe extract this item to the variables for better readability?

**Review by Cjkjvfnby — APPROVED** [review]

**Comment by Cjkjvfnby** [comment]
@Grummel7 changes look good. My comment is optional, feel free to merge PR.

File: default/python/AI/ResourcesAI.py (hunk 5 above)
Comment (Grummel7) [reply default/python/AI/ResourcesAI.py:172]: The only reason, this is hard to read at the moment, is that the indexes are numbers.
I'd rather make it more readable when doing the big refactoring.

**Pull request merged**